### ISSUE TITLE
Angular 2 support ?

### ISSUE THREAD

**Issue by mmc41** [json-schema-form/angular-schema-form#774]
Now that Angular2 has been released, is there any plans to port angular-schema-form to Angular 2 ?

**Comment by Anthropic** [comment]
Absolutely, we are already looking into it. Hard to give a timeline, but we have a few avenues we are looking at on the best way to proceed.

**Comment by mmc41** [comment]
@Anthropic  Ok. I guess I need to roll my own solution than of my Angular2 app (in fact already stated) I only need to support a subset of the functionality with a custom non-bootstrap gui anyway + I am using typescript 2.0 (not plain js). If it makes sense, we could work together but properly not easy to align requirements, time, architecture-ideas etc.

**Comment by jazzmanro** [comment]
Or, could a hybrid Angular 1 + 2 model be used meanwhile? We already use a v2 application and I guess that Angular's UpgradeAdapter may work to integrate v1 directives like schema form.
Does anyone have a working simple example how this could be achieved?

**Comment by Anthropic** [comment]
@mmc41 @jazzmanro I have already heard of people using the upgrade adapter to use it within an ng2 app.

**Comment by jazzmanro** [comment]
Would be so great if anyone could post a simple working example, even if it uses other library than schema form. My lack of Angular experience gave me too many headaches when trying this, without success so far.

**Comment by bombompb** [comment]
I agree with the fact that we would like to have a working example which uses the hybrid Angular 1-2 UpgradeAdapter till this project gets ported to Angular 2 completely. Can anyone help in this?

@Anthropic: Can you please try to give some estimate on the porting? How is it progressing?

**Comment by Anthropic** [comment]
@nicklasb did you know of a hybrid version working and how ng2 would work with the form?

**Comment by mmc41** [comment]
Nb. See also "https://github.com/json-schema-form/json-schema-form-core/issues/3" for a suggestion on how to more easily retarget for various web frameworks, incl. Angular 2.

**Comment by cheprus** [comment]
Here is probably a temporary alternative https://github.com/makinacorpus/angular2-schema-form
Untill the angular-schema-form will be ported to Angular 2.

**Comment by nicklasb** [comment]
@Anthropic , @mmc41 , @cheprus :
It is possible to use ASF in Angular 2 with ngUpgrade for the time being.

I use it in the administrative interface for my framework:
https://github.com/OptimalBPM/of-admin/wiki/App-Entry-&-Bootstrapping

**Comment by nicklasb** [comment]
The admin UI is also Angular 2, BTW.
Because A2 is about components, that part is kind of contained anyway, so it works well. 
Sure, if one starts to intertwine stuff directly, one might encounter some problems, on the other hand, that is not a very nice architectural design choice.

**Comment by YoYo35** [comment]
Hi
Any estimation for this issue?

**Comment by cmartin81** [comment]
+1

**Comment by Anthropic** [comment]
The first alpha of the core is out, once a few more fixes and a clean API are added then ng2 work can start using it. Timelines are a bit tricky to indicate, never know how much time anyone will have. I suspect once I get the ASF alpha out shortly that there may be a few more contributors willing to help work on it.

**Comment by mathpaquette** [comment]
what's the status for this issue. I'm going to have spare time in the near future and I wish to help.

**Comment by Anthropic** [comment]
@mathpaquette (and anyone interested in helping) can message me on Gitter.
I will make a Gitter group tonight for people willing to help with progressing this.

**Comment by mathpaquette** [comment]
Hi @Anthropic where's the Gitter? Also a thing that makes me very sad is that jsonschema form as been ported to recent framework like react (react-jsonschema-form) but nobody have considered to create an intermediate lib that we can easily reuse (probably its more efforts). I saw that you created a core project but is there a real interest by the community?

**Comment by Anthropic** [comment]
@mathpaquette yes there is likely to be an effort shortly to fork and port the react lib you mentioned over to use the core from here, some users want a react port that is 100% compatible using the core. I will work with them on that to ensure the core is as stable and usable as possible for any library. Gitter is linked from this repo home page but you can just go onto gitter and use my github username to find me anyway.

**Comment by GFoley83** [comment]
@Anthropic just wondering if there's any update or progress on Angular 2+ support? Have an Angular 4 enterprise project about to kick off and ASF would make life a million times easier. Thanks!

**Comment by Anthropic** [comment]
I don't expect it to be any time too soon, I have been sick for three weeks and taken off anything form related at work, so I am struggling on time, I have one colleague at work testing several different approaches for me though and I am working on making validator selectable to remove dependence on tv4, so it isn't a stand still, Angular no longer has a $compile method so we're determining the best way to overcome that at the moment since it is used heavily in the current version. Anyway using ngupgrade is still the main option for the moment.

**Comment by swirlsky** [comment]
@GFoley83 same here.

**Comment by dmastag** [comment]
Hi @Anthropic ,
Is there some progress on the Angular 4 implementation? Am looking really forward to it.

**Comment by Anthropic** [comment]
@dmastag I am working on it today, there's been a break as I was sick for four weeks over the last two months with two separate cases of the flu, but I am started back into it yesterday. Timelines are hard to provide though as it is only me working on it in spare time.

**Comment by dmastag** [comment]
Hi @Anthropic ,
Thanks for the update. Hope that you are feeling better now.
Is the work done on this repository? And is it still a huge task to do?

Maybe if the tasks can be split up into smaller jobs then the community is willing to help you on it?

**Comment by beeing** [comment]
Hi, just to share in case you may not know about this project - https://github.com/dschnelldavis/angular2-json-schema-form. Perhaps can merge the effort to improve on one library.

**Comment by GFoley83** [comment]
https://github.com/dschnelldavis/angular2-json-schema-form is still in alpha and development is pretty sparse. I opened an issue over a month ago asking if the project was still in active development and have yet to get a response.

I've resorted to using:
https://github.com/makinacorpus/angular2-schema-form

which is stable, actively in development and seem to be doing the trick so far.

**Comment by dmastag** [comment]
@GFoley83 agree, and it is even mentioned on the disclaimer that they think this is a great library and would like to join forces on an Angular2+ solution

https://github.com/makinacorpus/angular2-schema-form#disclaimer

**Comment by Anthropic** [comment]
Yes I have spoken to them to get them to add the disclaimer (people were getting confused by the name and asking me support questions for it) and we will probably collaborate together in future once the core work is done. I'll give you a list later about what work there is to do, I am always happy to help get someone up to speed to work on the project if they are interested I've trained a few people already, but they all ran out of bandwidth to help before being able to get involved. It is hard while transitioning to do docs as they'd become obsolete after a few weeks at this stage, but they will come.

**Comment by Anthropic** [comment]
@dmastag I'm currently working on:
 -  finish streamlining the build process in the core and making it TypeScript using build based on rxjs's build process, which I am almost finished with (almost is not referring to time, but effort).
 -  changing the validator in the core, tv4 is no longer supported and I want features from draft 6 and 7, I am working on making any validator supportable starting with djv and ajv.  
 -  They're the hardest parts, then I have to make AngularJS version pass all its tests again and document anything not backwards compatible.
 -  Then migrate AngularJS version of ASF to TypeScript
 -  Then migrate TS AngularJS version to TS Angular
 -  Then update to include any json-schema features that we don't currently support properly or from the later drafts.

I'm always willing to provide support to anyone who wants to be a co-maintainer to teach them how to work on it, but I suspect it will be much easier and more likely once things are moved to TypeScript and it becomes a lot easier to understand, debug and build.

**Comment by cmartin81** [comment]
Any updates here?

**Comment by srihari-sridharan** [comment]
Hi any updates on this?

**Comment by Anthropic** [comment]
No, I've been on holidays for a month+ and the end of year workload was too hectic, so I will be starting to work on ASF again this week. I have Assimilate and core running tests, but need to work on ASF to get it to compile and run tests properly so I can be sure it is working.

**Comment by srihari-sridharan** [comment]
Thanks for the update @Anthropic

**Comment by srihari-sridharan** [comment]
@Anthropic I found [angular2-schema-form](https://github.com/makinacorpus/angular2-schema-form). Asking this since I am new to Schema Forms, there is a disclaimer that states "We think angular-schema-form is a great Angular 1 library, and when it will move to Angular 2, we will probably join our efforts to produce and maintain a unique Angular 2 solution." Are there any plans on your end to join their effort? Adding @makinacorpus.

**Comment by Anthropic** [comment]
Yes I have had conversations with them when I asked them to put in the disclaimer and I hope to work with them once the core is stable, I've also talked with @dschnelldavis who put together https://github.com/dschnelldavis/angular2-json-schema-form

**Comment by srihari-sridharan** [comment]
Great to hear that! Please keep me posted as we are looking forward to use
this.



Thanks and Regards,

*Srihari Sridharan*

Never postpone work which should be done without delaying. Tirukkuṛaḷ
<http://en.wikipedia.org/wiki/Tirukkural> *2nd ** century BC*

தூங்குக தூங்கிச் செயற்பால் தூங்கற்க

தூங்காது செய்யும் வினை. - திருக்குறள்
<http://en.wikipedia.org/wiki/Tirukkural>

**Comment by srihari-sridharan** [comment]
@Anthropic can you please suggest me when to expect the support of Angular2? Asking this since my team is looking forward to using this. Thanks in advance.

**Comment by Anthropic** [comment]
@srihari-sridharan it all depends on how much support I get. Otherwise it is just me on my own so timelines are very difficult to predict, I had expected second half of last year, but my workload increased too much to make that possible, definitely not weeks though.

**Comment by Anthropic** [comment]
**I'll update this ticket whenever there are any updates.** I'm still time poor unfortunately, but things are slightly improving, however I am about to have an operation (nothing serious) that will see me offline for a few weeks.

I'm hoping to put out an update to **0.x** very soon with AngularJS 1.7 compatibility fixes and then finish off **1.x** in the next month after the operation. I then want to converge into a monorepo to make it easier for other developers to help in future on a major upgrade to **2.x** with **Angular support**. As always these plans and timelines are beholden to the whims of reality.

**Comment by anuj9196** [comment]
Any update on Angular 10+ support?
